### PR TITLE
SMTChecker tests and support for literal suffixes

### PR DESCRIPTION
**Pull request by matheusaaguiar**
task of #12656 .
Added tests and also ~a warning about missing~ support for literal suffix functions with two arguments.

EDIT: Support for functions with two arguments was indeed implemented.

### CODE REVIEW

File: libsolidity/formal/BMC.cpp
Comment (matheusaaguiar) [inline libsolidity/formal/BMC.cpp]: I didn't see a good place to create a function that does this and eliminates the code duplication, so I just left it for now.

Comment (leonardoalt) [reply libsolidity/formal/BMC.cpp]: Both modified functions call `SMTEncoder::endVisit(FunctionDefinition)`, so you could put it there instead and I think the behavior should be the same.
Also it should use `m_unsupportedErrors` instead of `m_errorReporter`.

File: test/libsolidity/smtCheckerTests/literalSuffixes/division_by_zero_function_single_argument.sol
Comment (leonardoalt) [inline test/libsolidity/smtCheckerTests/literalSuffixes/division_by_zero_function_single_argument.sol:8]: The new warning is not reported though

Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/literalSuffixes/division_by_zero_function_single_argument.sol:8]: Ah ok, new warning is only for two arguments. Is the suffix function actually properly supported then?

File: test/libsolidity/smtCheckerTests/literalSuffixes/division_by_zero_function_two_arguments.sol
Comment (leonardoalt) [inline test/libsolidity/smtCheckerTests/literalSuffixes/division_by_zero_function_two_arguments.sol]: It would be good to avoid this.

File: test/libsolidity/smtCheckerTests/literalSuffixes/equivalent_function_call_with_single_argument.sol
Comment (leonardoalt) [inline test/libsolidity/smtCheckerTests/literalSuffixes/equivalent_function_call_with_single_argument.sol:44]: Oh wow, looks like the functions are indeed supported?

File: test/libsolidity/smtCheckerTests/literalSuffixes/equivalent_function_call_with_two_arguments.sol
Comment (leonardoalt) [inline test/libsolidity/smtCheckerTests/literalSuffixes/equivalent_function_call_with_two_arguments.sol]: Nice!

Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/literalSuffixes/equivalent_function_call_with_two_arguments.sol]: Wait, what verification condition was proved safe here? The assertion fails above, so this would probably be the overflow check in `m + e`, but that should fail, shouldn't it?

File: test/libsolidity/smtCheckerTests/literalSuffixes/overflow_function_single_argument.sol
Comment (leonardoalt) [inline test/libsolidity/smtCheckerTests/literalSuffixes/overflow_function_single_argument.sol:25]: This is probably the underflow check for `intSuffix`, but we should verify.

File: test/libsolidity/smtCheckerTests/literalSuffixes/equivalent_function_call_with_single_argument.sol
Comment (cameel) [reply test/libsolidity/smtCheckerTests/literalSuffixes/equivalent_function_call_with_single_argument.sol:44]: Yeah, they look exactly like function calls in the AST. The only difference is the `isSuffixCall` annotation but that does not affect semantics (as long as the call passed parser and analysis). The ones with a single argument work exactly like functions. Only the ones with two need to account for the decomposition into mantissa and exponent so I think that will be the only case you'll need explicit handling for.

File: test/libsolidity/smtCheckerTests/literalSuffixes/equivalent_function_call_with_two_arguments.sol
Comment (matheusaaguiar) [reply test/libsolidity/smtCheckerTests/literalSuffixes/equivalent_function_call_with_two_arguments.sol]: Hum, not sure, but I am guessing that it doesn't fail because `123 + 2` is not overflow (?)

File: test/libsolidity/smtCheckerTests/literalSuffixes/overflow_function_single_argument.sol
Comment (matheusaaguiar) [reply test/libsolidity/smtCheckerTests/literalSuffixes/overflow_function_single_argument.sol:25]: How can one do that? Some warnings mention an option `show proved safe`, but I haven't figured out how to enable that via `isoltest`.

File: test/libsolidity/smtCheckerTests/literalSuffixes/equivalent_function_call_with_two_arguments.sol
Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/literalSuffixes/equivalent_function_call_with_two_arguments.sol]: Yea but that's just one case. There's also the general case `m + e` inside the function.

Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/literalSuffixes/equivalent_function_call_with_two_arguments.sol]: Ahh function `suffix` is internal though, right?

File: test/libsolidity/smtCheckerTests/literalSuffixes/overflow_function_single_argument.sol
Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/literalSuffixes/overflow_function_single_argument.sol:25]: Via `isoltest` that won't work because the settings are specific to the SMTCheckerTests. You'd need to run `solc` itself with `--model-checker-engine all --model-checker-show-proved-targets` or smth like that

File: test/libsolidity/smtCheckerTests/literalSuffixes/equivalent_function_call_with_two_arguments.sol
Comment (cameel) [reply test/libsolidity/smtCheckerTests/literalSuffixes/equivalent_function_call_with_two_arguments.sol]: Yes, suffixes are always internal calls.

Comment (matheusaaguiar) [reply test/libsolidity/smtCheckerTests/literalSuffixes/equivalent_function_call_with_two_arguments.sol]: Yes, suffix function calls are internal.

File: test/libsolidity/smtCheckerTests/literalSuffixes/overflow_function_single_argument.sol
Comment (matheusaaguiar) [reply test/libsolidity/smtCheckerTests/literalSuffixes/overflow_function_single_argument.sol:25]: Thanks! 
The option is  `--model-checker-show-proved-safe`.

File: test/libsolidity/smtCheckerTests/literalSuffixes/equivalent_function_call_with_two_arguments.sol
Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/literalSuffixes/equivalent_function_call_with_two_arguments.sol]: Ok, then yea that property is proved safe because it's only reachable with `123 + 2`.

File: libsolidity/formal/BMC.cpp
Comment (matheusaaguiar) [reply libsolidity/formal/BMC.cpp]: I tried that at first, but, unfortunately, `CHC::endVisit` does not call `SMTEncoder::endVisit(FunctionCall)` for suffix functions which are internal.
Instead, we have the following call sequence: `CHC::internalFunctionCall` -> `CHC::predicate` -> `SMTEncoder::symbolicArguments`. 
That ends in an [assertion fail](https://github.com/ethereum/solidity/blob/fa744abf623bbb5ac02d32a99711ebfb1c7d37d5/libsolidity/formal/SMTEncoder.cpp#L3177), because as @cameel explained in another message, two argument suffix functions need to decompose the literal into mantissa and exponent.
 I was thinking more of an already existing namespace/class that could house such function until SMTChecker properly support such suffix functions.

Comment (matheusaaguiar) [reply libsolidity/formal/BMC.cpp]: Maybe creating a `checkAndWarnUnsupportedSuffixFunction` in `SMTEncoder`?

**Comment by cameel** [comment]
I pushed new code to the base branch. Please rebase.

File: libsolidity/formal/BMC.cpp
Comment (cameel) [inline libsolidity/formal/BMC.cpp]: I'm not sure this is true. SMTChecker will properly see this as a function call in the AST and won't mistake it for a literal without a suffix. The thing that is wrong is that it will see one argument passed into the call while the function accepts two arguments.

I'm not sure what it will actually do. I'd actually expect an assertion failure somewhere down the line.

Comment (cameel) [reply libsolidity/formal/BMC.cpp]: If that happens, you should find the place that fails and make it ignore the call (or do something else that's possibly incorrect but won't make it crash) when  `_funCall.isSuffixCall()` is true. You should also add a FIXME there to make it clear it's just a hack to make it not crash.

File: libsolidity/formal/BMC.cpp
Comment (cameel) [reply libsolidity/formal/BMC.cpp]: I would not worry too much about the warning being duplicated. It's temporary anyway.

Not sure about `SMTEncoder::endVisit(FunctionCall)`, I'd have to look deeper into the code, but one thing I can suggest is that you should find the failing assertion and bypass it. Or you could calculate the arguments that's wrong but simple - e.g. set the first one the the value of the literal and the second one to zero.

File: test/libsolidity/smtCheckerTests/literalSuffixes/division_by_zero_function_two_arguments.sol
Comment (cameel) [reply test/libsolidity/smtCheckerTests/literalSuffixes/division_by_zero_function_two_arguments.sol]: This is probably the disconnect between the function call and the definition. I.e. we have only one argument in the `FunctionCall` in the AST (`1234567890`) but the function requires two. When the SMTChecker encounters this it should call `RationalNumberType::fractionalDecomposition()` to split it and treat the result as actual arguments.

Doing this would probably let us actually get rid of the warning because it's all that's needed to support those suffixes.

File: libsolidity/formal/BMC.cpp
Comment (cameel) [reply libsolidity/formal/BMC.cpp]: Actually, like I said in another comment, you can probably just do the fractional decomposition (which is just a single function call) and get rid of the warning because that will make the suffixes work properly.

Comment (leonardoalt) [reply libsolidity/formal/BMC.cpp]: Is that the only case with > 1 arguments?

Comment (cameel) [reply libsolidity/formal/BMC.cpp]: Yes. Suffixes are basically function calls. The need to split rational literals to handle numbers like 1.23 is the only codegen-level difference. There are other differences but only at analysis level. They are only restricting what you can do compared to normal function calls, but not adding new capabilities.

Comment (leonardoalt) [reply libsolidity/formal/BMC.cpp]: Ok yea, then it would be probably easy enough to actually fully support it.

Comment (matheusaaguiar) [reply libsolidity/formal/BMC.cpp]: I think I did that in https://github.com/ethereum/solidity/pull/14118/commits/2ea0c9dcbf13702255708741175ce26c661d867f, but I am not sure if that is properly implemented, since I just created the expressions "on the fly" and did not add them to the `context`.

https://github.com/ethereum/solidity/blob/2ea0c9dcbf13702255708741175ce26c661d867f/libsolidity/formal/SMTEncoder.cpp#L3177-L3185

File: libsolidity/formal/SMTEncoder.cpp
Comment (cameel) [inline libsolidity/formal/SMTEncoder.cpp]: I'd add some asserts here:
```solidity
solAssert(!funType->hasBoundFirstArgument());
solAssert(arguments.size() == 1);
```

Comment (cameel) [inline libsolidity/formal/SMTEncoder.cpp]:
```suggestion
		auto const* literalType = dynamic_cast<RationalNumberType const*>(arguments[0]->annotation().type);
```
or
```suggestion
		auto&& literalType = dynamic_cast<RationalNumberType const*>(arguments[0]->annotation().type);
```

File: libsolidity/formal/SMTEncoder.cpp
Comment (cameel) [inline libsolidity/formal/SMTEncoder.cpp]:
```suggestion
		args.emplace_back(mantissa->literalValue(nullptr));
		args.emplace_back(exponent->literalValue(nullptr));
```

Comment (cameel) [inline libsolidity/formal/SMTEncoder.cpp]: Looks like negative literals require special handling (as 2's complement via `u2s()`) so you may also want to assert that mantissan and exponent are not negative (with `RationalNumberLiteral::isNegative()`).

File: test/libsolidity/smtCheckerTests/literalSuffixes/equivalent_function_call_with_two_arguments.sol
Comment (cameel) [inline test/libsolidity/smtCheckerTests/literalSuffixes/equivalent_function_call_with_two_arguments.sol:7]: Let's add a case for an integer too:
```suggestion
        assert(1.23 suffix == suffix(123, 2));
        assert(123 suffix == suffix(123, 0));
        assert(0 suffix == suffix(0, 0));
```

File: test/libsolidity/smtCheckerTests/literalSuffixes/overflow_function_single_argument.sol
Comment (cameel) [inline test/libsolidity/smtCheckerTests/literalSuffixes/overflow_function_single_argument.sol:1]: I guess we should have overflow/underflow tests for two argument functions for completeness as well?

**Review by cameel — COMMENTED** [review]
Needs some asserts and small tweaks but other than that looks like it works!

File: libsolidity/formal/BMC.cpp
Comment (cameel) [reply libsolidity/formal/BMC.cpp]: Well, seems that this is what `constantExpr()` does too. It returns the value from `literalValue()` directly. And here the situation is even simpler because the decomposed arguments are not even expressions in the AST. They're just some constant numbers we get based on the constant input we have.

It looks to me like it should work but of course it's best if @leonardoalt can confirm.

File: test/libsolidity/smtCheckerTests/literalSuffixes/overflow_function_single_argument.sol
Comment (matheusaaguiar) [reply test/libsolidity/smtCheckerTests/literalSuffixes/overflow_function_single_argument.sol:1]: Yes, I will add those now that we got support for them.

File: libsolidity/formal/SMTEncoder.cpp
Comment (cameel) [inline libsolidity/formal/SMTEncoder.cpp]: The point of `emplace_back()` was to get rid of the explicit conversion:
```suggestion
		args.emplace_back(mantissa->literalValue(nullptr));
		args.emplace_back(exponent->literalValue(nullptr));
```

File: libsolidity/formal/SMTEncoder.cpp
Comment (cameel) [inline libsolidity/formal/SMTEncoder.cpp]: Let's actually go with the universal reference here:
```suggestion
		auto&& literalType = dynamic_cast<RationalNumberType const*>(arguments[0]->annotation().type);
```

I'm switching to that style in the main PR as well.

File: libsolidity/formal/SMTEncoder.cpp
Comment (matheusaaguiar) [reply libsolidity/formal/SMTEncoder.cpp]: Ah yes, I mindlessly replaced `push_back()` and forgot about that.

File: test/libsolidity/smtCheckerTests/literalSuffixes/overflow_function_single_argument.sol
Comment (cameel) [inline test/libsolidity/smtCheckerTests/literalSuffixes/overflow_function_single_argument.sol]: `z` is not a constant :)

**Review by cameel — APPROVED** [review]
Looks good. Please squash the commits.

**Comment by cameel** [comment]
Now waiting for approval from @leonardoalt and @ekpyron. Then we can merge it into the main PR (or `develop` if the main PR gets merged before).

File: test/libsolidity/smtCheckerTests/literalSuffixes/overflow_function_single_argument.sol
Comment (matheusaaguiar) [reply test/libsolidity/smtCheckerTests/literalSuffixes/overflow_function_single_argument.sol]: Yeah, fair enough :)

**Review by cameel — APPROVED** [review]

File: test/libsolidity/smtCheckerTests/literalSuffixes/equivalent_function_call_with_two_arguments.sol
Comment (leonardoalt) [inline test/libsolidity/smtCheckerTests/literalSuffixes/equivalent_function_call_with_two_arguments.sol:10]: Can you also add some assertions that should fail?

**Review by leonardoalt — APPROVED** [review]
Asked for one more test, but also fine with it as is.

File: test/libsolidity/smtCheckerTests/literalSuffixes/equivalent_function_call_with_two_arguments.sol
Comment (matheusaaguiar) [reply test/libsolidity/smtCheckerTests/literalSuffixes/equivalent_function_call_with_two_arguments.sol:10]: I added them in a separate file `equivalent_function_call_with_two_arguments_fail.sol`, ok?

**Comment by cameel** [comment]
We decided not to merge this feature so I'm closing the PR. See https://github.com/ethereum/solidity/issues/13718#issuecomment-1551488965 for details.

**Pull request closed without merging**
